part/part_test.go

```diff
@@ -0,0 +1,544 @@
+	default:
+	}
+
+	_, _, tree = tree.Insert([]byte("a"), []byte("b"))
+
+	select {
+	case <-watch:
+	case <-time.After(10 * time.Second):
+		t.Fatal("expected watch channel to close")
+	}
+
+	v, _, ok := tree.Get([]byte("a"))
+	if !ok {
+		t.Fatal("expected to find 'a'")
+	}
+	if string(v) != "b" {
+		t.Fatal("expected value 'b'")
+	}
+}
+
+func Test_insert(t *testing.T) {
+	tree := New[int]()
+	for i := 0; i < 10000; i++ {
+		key := binary.NativeEndian.AppendUint32(nil, uint32(i))
+		_, _, tree = tree.Insert(key, i)
+	}
+	for i := 0; i < 10000; i++ {
+		key := binary.NativeEndian.AppendUint32(nil, uint32(i))
+		_, _, ok := tree.Get(key)
+		if !ok {
+			t.Fatalf("%d not found", i)
+		}
+	}
+}
+
+func Test_prefix(t *testing.T) {
+	tree := New[[]byte]()
+	ins := func(s string) { _, _, tree = tree.Insert([]byte(s), []byte(s)) }
+	ins("a")
+	ins("ab")
+	ins("abc")
+	ins("abcd")
+
+	iter, _ := tree.Prefix([]byte("ab"))
+	k, v, ok := iter.Next()
+	assert.True(t, ok)
+	assert.Equal(t, []byte("ab"), k)
+	assert.Equal(t, []byte("ab"), v)
+
+	k, v, ok = iter.Next()
+	assert.True(t, ok)
+	assert.Equal(t, []byte("abc"), k)
+	assert.Equal(t, []byte("abc"), v)
+
+	k, v, ok = iter.Next()
+	assert.True(t, ok)
+	assert.Equal(t, []byte("abcd"), k)
+	assert.Equal(t, []byte("abcd"), v)
+
+	_, _, ok = iter.Next()
+	assert.False(t, ok)
+}
+
+func Test_txn(t *testing.T) {
+	tree := New[uint64]()
+	ins := func(n uint64) { _, _, tree = tree.Insert(intKey(n), n) }
+
+	var iter *Iterator[uint64]
+	next := func(exOK bool, exVal int) {
+		t.Helper()
+		_, v, ok := iter.Next()
+		if assert.Equal(t, exOK, ok) {
+			assert.EqualValues(t, exVal, v)
+		}
+	}
+
+	for i := 1; i <= 3; i++ {
+		ins(1)
+		ins(2)
+		ins(3)
+	}
+
+	txn := tree.Txn()
+	txn.Delete(intKey(2))
+	txn.Delete(intKey(3))
+	txn.Insert(intKey(4), 4)
+
+	iter = txn.Iterator()
+	next(true, 1)
+	next(true, 4)
+	next(false, 0)
+
+	_ = txn.Commit() // Ignore the new tree
+
+	// Original tree should be untouched.
+	for i := 1; i <= 3; i++ {
+		_, _, ok := tree.Get(intKey(uint64(i)))
+		assert.True(t, ok, "Get(%d)", i)
+	}
+
+	iter = tree.Iterator()
+	next(true, 1)
+	next(true, 2)
+	next(true, 3)
+	next(false, 0)
+}
+
+func Test_lowerbound(t *testing.T) {
+	tree := New[uint64]()
+	ins := func(n int) { _, _, tree = tree.Insert(intKey(uint64(n)), uint64(n)) }
+
+	// Insert 1..3
+	for i := 1; i <= 3; i++ {
+		ins(i)
+	}
+
+	var iter *Iterator[uint64]
+	next := func(exOK bool, exVal int) {
+		t.Helper()
+		_, v, ok := iter.Next()
+		require.Equal(t, exOK, ok)
+		require.EqualValues(t, exVal, v)
+	}
+
+	iter = tree.LowerBound([]byte{})
+	next(true, 1)
+	next(true, 2)
+	next(true, 3)
+	next(false, 0)
+
+	iter = tree.LowerBound(intKey(0))
+	next(true, 1)
+	next(true, 2)
+	next(true, 3)
+	next(false, 0)
+
+	iter = tree.LowerBound(intKey(3))
+	next(true, 3)
+	next(false, 0)
+
+	iter = tree.LowerBound(intKey(4))
+	next(false, 0)
+}
+
+func Test_iterate(t *testing.T) {
+	sizes := []int{0, 1, 10, 100, 1000, rand.Intn(1000)}
```

Comment:
any specific behaviour we want for calling Next() on iterators which already finished?

Reply:
Added assert.